Add option to specify additional cluster role for workspaces.

johnmcollier opened the pull request:
Signed-off-by: John Collier <John.J.Collier@ibm.com>

### What does this PR do?
Adds an option to specify an additional cluster role for workspaces. When the `CHE_WORKSPACE_CLUSTER_ROLE` value is set, Che will create an additional role binding during workspace creation that references the specified cluster role. 

This allows the user to create/use workspaces that require further roles than are provided by default (such as sidecar that needs to list/edit ConfigMaps, creating a service, etc).


### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/12972

#### Release Notes
Adds an option to specify an optional cluster role for workspaces

eivantsov commented:
Looks ok to me. Thanks @johnmcollier. The only thing is that che service account should be able to get ClusterRoles, otherwise it will fails to create such a roleBinding. But since this env is optional, ie an admin knows what he's doing, that's ok.

@sleshchenko wdyt?

johnmcollier commented:
@eivantsov That's a fair point. I could add a check when it goes to create a rolebinding, and if it fails, it will log a message.

sleshchenko commented on deploy/kubernetes/helm/che/templates/configmap.yaml:
```diff
@@ -101,3 +104,4 @@ data:
   JAEGER_SAMPLER_PARAM: "1"
   JAEGER_REPORTER_MAX_QUEUE_SIZE: "10000"
   CHE_METRICS_ENABLED: {{ .Values.global.metricsEnabled | quote }}
+  
```
Please make sure that this file has only one empty line in the end.

sleshchenko commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesWorkspaceServiceAccount.java:
```diff
@@ -82,6 +82,17 @@ void prepare() throws InfrastructureException {
+
+    // If the user specified an additional cluster role for the workspace,
+    // create a role binding for it too
+    String customClusterRoleName = System.getenv("CHE_WORKSPACE_CLUSTER_ROLE");
```
it would be better to do the same as with service account name. I mean use configuration property instead of env var directly.
See https://github.com/eclipse/che/blob/7b6cd44e36e94d40f366b5dcd2fe46074e8a7bad/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java#L40

sleshchenko replied:
P.S. `che.infra.kubernetes.cluster_service_account_name` would be right configuration property name, I guess. Also, add the default value to `che.properties` file

johnmcollier replied:
@sleshchenko Quick question: If both the environment variable and the Che property are set, which should be used? Or should I remove the environment variable and only use the Che property?

johnmcollier commented:
@sleshchenko Thanks for the review! I've updated the PR with the changes:
- Extra blank line from the configmap has been removed
- Created the `che.infra.kubernetes.cluster_role_name` property, and added it to the Che.properties file. Default value is NULL.
- Updated my changes to pull the value from `che.infra.kubernetes.cluster_role_name` instead of from the environment variable.

amisevsk commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesWorkspaceServiceAccount.java:
```diff
@@ -82,6 +87,16 @@ void prepare() throws InfrastructureException {
+          .rbac()
+          .kubernetesRoleBindings()
+          .inNamespace(namespace)
+          .createOrReplace(createCustomRoleBinding(this.clusterRoleName));
```
What does an error look like when `this.clusterRoleName` does not exist on the cluster? I'm mostly worried about troubleshooting in case of a typo.

johnmcollier replied:
@amisevsk Good point! Let me add a quick log message in the event the clsuter role doesn't exist.

sleshchenko commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesWorkspaceServiceAccount.java:
```diff
@@ -34,15 +38,21 @@
   private final String workspaceId;
+  private final String clusterRoleName;
+
+  private static final Logger LOG =
```
Please place the static field above non-static fields.

sleshchenko commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesWorkspaceServiceAccount.java:
```diff
@@ -165,4 +191,23 @@ private KubernetesRoleBinding createExecRoleBinding() {
         .build();
   }
+
+  private KubernetesRoleBinding createCustomRoleBinding(String clusterRoleName) {
```
Should not `cluster` be used here instead of `custom`?

johnmcollier replied:
I decided to use `createCustomRoleBinding` since we're creating a role binding for the cluster role here. 

Note that role bindings can bind both Roles and ClusterRoles

sleshchenko commented:
ci-build

sleshchenko commented:
ci-test

johnmcollier commented:
@sleshchenko Thanks again for your review! I've moved the static logger to above the non-static fields.

sleshchenko commented:
ci-build

musienko-maxim commented:
Can one of the admins verify this PR?

sleshchenko commented:
ci-build

johnmcollier commented:
My branch was a little out of date, so I updated it against master.

metlos commented:
ci-build

johnmcollier commented:
ci-build

johnmcollier commented:
@sleshchenko @metlos Since I don't have permission to trigger a new CI build, would one of you guys be able to kick a new build off? Thanks!

sleshchenko commented:
ci-build

johnmcollier commented:
@sleshchenko @metlos Sorry to ping again! Could you guys also run `ci-test` if needed? And assuming the tests are good, is there anything else I need to do before it's merged? Thanks!

eivantsov commented:
ci-test

johnmcollier commented:
All tests passed 😄 
<img width="1461" alt="Screen Shot 2019-04-24 at 3 53 53 PM" src="https://user-images.githubusercontent.com/6880023/56689537-4c78c480-66a9-11e9-85a3-a88b5f2fbd9f.png">

johnmcollier commented:
Hey, just wondering if the PR is good to be merged now?
